hazelcast-jet-core/src/main/java/com/hazelcast/jet/impl/pipeline/Planner.java

```diff
@@ -111,6 +113,11 @@ DAG createDag() {
             }
         }
 
+        // determine exact local parallelism of the transforms
+        for (Transform transform : adjacencyMap.keySet()) {
+            transform.determineLocalParallelism(ctx);
```

Comment:
Here we mutate the pipeline. The following code will have unexpected behavior.
```java
Pipeline p = ...
DAG dagWithLp1 = p.toDag(() -> 1);
DAG dagWithLp2 = p.toDag(() -> 2);
```
Both DAGs will use LP=1 where default LP was used.

I also don't see a reason why do we need this. Can't we have the vertices with LP=1 and rely on the logic in `Vertex` class?

Reply:
You mean -1?  We have to propagate the parallelism from source to sink, i don't think that's something a `Vertex` should do.

As for mutating the pipeline, I think that's fine because it happens on the server side, on a deserialized copy.

I think we are also storing the pipeline with the job, for this it would be important to serialize it before modification.

Reply:
>As for mutating the pipeline, I think that's fine because it happens on the server side, on a deserialized copy

You still can call `pipeline.toDag()` on the client. For what reason I don't know, but you can. It might not be an issue in practice, but it's not pure and it can be very surprising if you hit it. Methods like `toDag` _shouldn't_ surprisingly change the local parallelism of all stages in the pipeline.

Reply:
>I think we are also storing the pipeline with the job, for this it would be important to serialize it before modification.

We always store the `DAG` in the `JobRecord`.

Reply:
> We always store the DAG in the JobRecord.

Jet could restart with a different default local parallelism, so this is going to become an issue.

Reply:
We assume all members have equal hardware and same configuration, so the local parallelism should be same...